Implement refreshing the Cloud token in Enso libraries

radeusgd opened the pull request:
### Pull Request Description

- Closes #9300 
- Now the Enso libraries are themselves capable of refreshing the access token, thus there is no more problems if the token expires during a long running workflow.
- Adds `get_optional_field` sibling to `get_required_field` for more unified parsing of JSON responses from the Cloud.
	- Adds `expected_type` that checks the type of extracted fields. This way, if the response is malformed we get a nice Enso Cloud error telling us what is wrong with the payload instead of a `Type_Error` later down the line.
- Fixes `Test.expect_panic_with` to actually catch only panics. Before it used to also handle dataflow errors - but these have `.should_fail_with` instead. We should distinguish these scenarios.

### Important Notes

<!--
- Mention important elements of the design.
- Mention any notable changes to APIs.
-->

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

radeusgd committed:
refactor: split AuthenticationProvider with CloudAPI, work on Authentication_Service that is automatically instantiated from Java

radeusgd committed:
fixes after refactor

radeusgd committed:
retry logic for fetching a secret

radeusgd committed:
logic when to refresh the token

radeusgd committed:
fixes

radeusgd committed:
panics

radeusgd committed:
remove empty file??

radeusgd committed:
WIP refresh flow

radeusgd committed:
add first test stub

radeusgd committed:
renames

radeusgd committed:
Not_Logged_In is a panic

radeusgd committed:
updating tests

radeusgd committed:
finish the logic for refreshing the token

radeusgd committed:
remove placeholder

radeusgd committed:
remove debug print

radeusgd committed:
improve Panic test catcher

radeusgd committed:
better test

radeusgd committed:
clearer error if token is expired

radeusgd committed:
fix in Test

radeusgd committed:
fix in Auth

radeusgd committed:
add test

radeusgd committed:
fix in Test 2

radeusgd committed:
fix to_display_text for Not_Logged_In

radeusgd committed:
token lifetime

radeusgd committed:
advanced credential mocking

radeusgd committed:
fixes

radeusgd committed:
add basic tests for refresh scenarios

radeusgd committed:
make secret read robust also to service errors

radeusgd committed:
fixes

radeusgd committed:
support 403 in secret

radeusgd committed:
fix panic test

radeusgd committed:
implementing token renewal logic in mock

radeusgd committed:
fixing tests

radeusgd committed:
implement retry logic

radeusgd committed:
add test to secrets

radeusgd committed:
doc

radeusgd committed:
un-indent

radeusgd committed:
count expired tokens to prove retries are happening

radeusgd commented on app/gui/view/graph-editor/src/builtin/visualization/java_script/table.js:
I removed this file because it was empty and it was causing problems on my machine.

That's why this PR has triggered the GUI codeowners.

radeusgd commented on std-bits/base/src/main/java/org/enso/base/enso_cloud/CloudAPI.java:
```diff
@@ -0,0 +1,42 @@
+      return null;
+    }
+
+    // TODO we should be able to fetch the name and organizationId from the cloud:
+    String directoryName = "???";
+    String organizationId = "";
+    cachedWorkingDirectory = new CloudWorkingDirectory(directoryName, directoryId, organizationId);
+    return cachedWorkingDirectory;
```
This is not new code, just moved during a refactor.
I guess it should be handled around #9289

radeusgd committed:
doc

radeusgd committed:
javafmt

radeusgd committed:
docs

GregoryTravis commented on distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Errors.enso:
```diff
@@ -5,10 +5,23 @@ import project.Network.HTTP.HTTP_Status_Code.HTTP_Status_Code
    Error if the user is not logged into Enso Cloud.
 type Not_Logged_In
+    ## PRIVATE
+    Error
```
Why is it preferred to have a constructor?

radeusgd replied:
Without the constructor the `to_display_text` method did not work correctly.

Singleton types are not specified well enough IMO in current Enso implementation and method dispatch on them is a bit wonky - I guess I could also remove the `self` argument from `to_display_text` to make it a 'static' method, but I was less sure that this solution will work right in all contexts, so it felt safer to just use the same pattern we use elsewhere - a regular atom type instead of singleton.

GregoryTravis commented on distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Internal/Authentication.enso:
```diff
@@ -0,0 +1,166 @@
+
+    ## PRIVATE
+    read_from_credentials -> Authentication_Data =
+        content = credentials_file.read_text . catch File_Error error-> case error of
+            File_Error.Not_Found _ -> Panic.throw Not_Logged_In.Error
+            _ -> Panic.throw (Illegal_State.Error "Failed to read credentials file: "+error.to_display_text cause=error)
+        as_json = content.parse_json.catch Invalid_JSON error->
+            Panic.throw (Illegal_State.Error "Invalid credentials file format: "+error.to_display_text cause=error)
+        get_field field_name =
+            as_json.get field_name if_missing=(Panic.throw (Illegal_State.Error "Invalid credentials file format: missing field `"+field_name+"`."))
```
This could be shared with the similar code in `Refresh_Token_Data`, passing in a prefix for error messages.

radeusgd replied:
IMO it's such a small piece of code that abstracting it and adding the prefix makes the code more complicated, I did not share it on purpose.

radeusgd replied:
After second thought I will try to improve the error handling a bit.

radeusgd replied:
Refactored in https://github.com/enso-org/enso/pull/9390/commits/d87232886ab733cc267d0f0ed55bc1cd980f2eb2

jdunkerley commented on distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Internal/Utils.enso:
```diff
@@ -86,42 +60,53 @@ datalinks_api = cloud_root_uri + "connectors"
+        if response.code.is_success then response else
+            is_unauthorized = (response.code == HTTP_Status_Code.unauthorized) || (response.code == HTTP_Status_Code.forbidden)
+            case is_unauthorized of
+                True -> if retries < 0 then Error.throw Enso_Cloud_Error.Unauthorized else
+                    Authentication.refresh_access_token
```
Feels wrong to run the refresh 3 times? Small detail but seems like should fail if unauthorized and has done the refresh token.

radeusgd replied:
What other approach would you suggest?

I don't see any simpler ones, but I'm open to suggestions.

Currently the retry logic retries up to 3 times regardless of the kind of error.

What if the 1st retry fails due to network issue, second needs token refresh but also fails due to 500 server error and only the 3rd one succeeds? I don't see a difference between retrying for different reasons.

Note that if the 'refresh token' operation itself fails - that has no retry so it will fail regardless of the retry count. Shall I add the retry there too maybe?

jdunkerley commented on app/gui/view/graph-editor/src/builtin/visualization/java_script/table.js:
why has this been touched here?

radeusgd replied:
https://github.com/enso-org/enso/pull/9390#discussion_r1522073920

farmaazon commented on app/gui/view/graph-editor/src/builtin/visualization/java_script/table.js:
Some leftover from the old GUI? It caught me as reviewer to this PR.

radeusgd replied:
That seems to be the case. [As mentioned](https://github.com/enso-org/enso/pull/9390#discussion_r1522073920) it was just an empty file that was causing compilation problems for me, so I removed it.

@farmaazon can you approve this change?

farmaazon replied:
Ah, so it is being removed from repo, not added (GH didn't display usual message "hiding because file is deleted" message, what confused me).

radeusgd committed:
CR: refactor and make the error handling even more robust

radeusgd committed:
fix in Test

radeusgd committed:
fix in engine

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/builtin/Error.java:
```diff
@@ -184,7 +184,7 @@ public Atom makeNoConversionCurrying(
   @CompilerDirectives.TruffleBoundary
   public Atom makeTypeError(Object expected, Object actual, String name) {
-    return typeError.newInstance(expected, actual, Text.create("`" + name + "`"));
+    return typeError.newInstance(expected, actual, Text.create("Expected `" + name + "` to be {exp}, but got {got}."));
```
Ajajaj, omission.

radeusgd committed:
javafmt

radeusgd committed:
Merge branch 'develop' into wip/radeusgd/9300-refresh-cloud-token

radeusgd committed:
better docs and refactor

radeusgd committed:
add expected types for better handling of malformed responses

radeusgd committed:
handle dataflow

radeusgd committed:
fix: the test was using `expect_panic_with`, but actually a dataflow error was being returned.

After I fixed `Test.expect_panic_with` to only intercept panics, this test started failing

radeusgd commented on test/Base_Tests/src/Data/Array_Proxy_Spec.enso:
```diff
@@ -43,7 +43,7 @@ add_specs suite_builder =
         group_builder.specify "should check validity at construction" <|
             Test.expect_panic_with (Array_Proxy.new 0 0) Type_Error
-            Test.expect_panic_with (Array_Proxy.new -1 (x->x)) Illegal_Argument
+            (Array_Proxy.new -1 (x->x)).should_fail_with Illegal_Argument
```
I have now fixed the `Test.expect_panic_with` to only intercept **actual Panics**. Before it used to happily catch dataflow errors as well and not distinguish between them.

Our tests should distinguish between the two cases.

After my changes this test used to fail with:
![image](https://github.com/enso-org/enso/assets/1436948/ea72cd33-64a0-4c21-ab26-c16e4c2a164c)

radeusgd committed:
wrong sig

radeusgd committed:
replace Regex helper with type assertion

radeusgd committed:
correct type errors usage

radeusgd committed:
remove obsolete import

radeusgd committed:
adapt tests to type errors always being a panic

radeusgd committed:
clearer handling of panic

radeusgd committed:
fix test not being run

radeusgd committed:
Merge branch 'develop' into wip/radeusgd/9300-refresh-cloud-token

radeusgd committed:
more tests not being run, fixes

radeusgd committed:
get rid of `should_fail_with Type_Error` in favour of `Test.expect_panic`

radeusgd committed:
missing import in micro-distribution

radeusgd committed:
Merge branch 'develop' into wip/radeusgd/9300-refresh-cloud-token

radeusgd committed:
yet another fix

radeusgd committed:
'polyfilling' `Text.replace` for micro-distribution rendering of Type Error that got more complicated recently

radeusgd committed:
dots

radeusgd committed:
Merge branch 'develop' into wip/radeusgd/9300-refresh-cloud-token

radeusgd committed:
fix

radeusgd committed:
Merge branch 'develop' into wip/radeusgd/9300-refresh-cloud-token

radeusgd committed:
Merge branch 'develop' into wip/radeusgd/9300-refresh-cloud-token
